File: packages/cli/test/fixtures/release-smoke/stripe-not-verified.ts
// Fixture: a Stripe webhook handler with the express.json() middleware-ordering bug.
// Phase 4.2 release-binaries.yml smoke step scans this directory and asserts the
// compiled binary recognises this as RULES-03 not-verified (express.json before the
// webhook route consumes the raw body Stripe needs for HMAC verification).
// Adapted from test/fixtures/precommit/stripe-not-verified.ts; same bug, fresh
// header so this file is not mistaken for the pre-commit fixture.
import express from "express";
import Stripe from "stripe";

const app = express();
const stripe = new Stripe(process.env.STRIPE_SECRET_KEY!, { apiVersion: "2024-06-20" });

// BUG: express.json() must NOT run before /webhook — it consumes the raw body Stripe needs.
app.use(express.json());

app.post("/webhook", (req, res) => {
  const sig = req.headers["stripe-signature"] as string;
  // BUG: req.body is the parsed JSON object, not the raw bytes — verification will fail silently
  // or constructEvent will throw. Either way, this handler is NOT VERIFIED.
  const event = stripe.webhooks.constructEvent(req.body, sig, process.env.STRIPE_WEBHOOK_SECRET!);
  res.status(200).json({ received: true, type: event.type });
});

app.listen(3000);

File: packages/cli/test/fixtures/release-smoke/stripe_not_verified.py
# Fixture: a Flask Stripe webhook handler with NO signature verification.
# Phase 4.2 release-binaries.yml smoke step scans this directory; the compiled
# binary must recognise this as RULES-02 not-verified (handler reads the body
# and parses JSON, never validates the Stripe-Signature header).
# Exercises the embedded WASM Python grammar on every target leg (DIST-05).
#
# This file deliberately omits the verification API and the standard-library
# crypto primitive a webhook handler would normally use, so the not-verified
# rule fires cleanly in the smoke step.
import json

from flask import Flask, jsonify, request

app = Flask(__name__)


@app.route("/webhook", methods=["POST"])
def webhook() -> tuple:
    payload = request.get_data()
    # BUG: signature header read but never checked against the payload.
    _sig_header = request.headers.get("Stripe-Signature", "")
    event = json.loads(payload)
    return jsonify({"received": True, "type": event.get("type")}), 200


if __name__ == "__main__":
    app.run(port=3000)
